bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4Handler.java

```diff
@@ -0,0 +1,852 @@
+        sendPacketEncrypted(outPacket, false);
+    }
+
+    private void turnOnPS4() {
+        wakeUpPS4();
+        scheduler.schedule(this::waitAndConnectToPS4, 13, TimeUnit.SECONDS);
+    }
+
+    private void waitAndConnectToPS4() {
+        try {
+            getConnection();
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+        }
+    }
+
+    private void sendStandby() {
+        ByteBuffer outPacket = PS4PacketHandler.makeStandbyPacket();
+        sendPacketEncrypted(outPacket);
+    }
+
+    /**
+     * Ask PS4 if the OSK is open so we can get and set text.
+     */
+    private void sendOSKStart() {
+        ByteBuffer outPacket = PS4PacketHandler.makeOSKStartPacket();
+        sendPacketEncrypted(outPacket);
+    }
+
+    /**
+     * Sets the entire OSK string on the PS4.
+     *
+     * @param text The text to set in the OSK.
+     */
+    private void setOSKText(String text) {
+        logger.debug("Sending osk text packet,\"{}\"", text);
+        ByteBuffer outPacket = PS4PacketHandler.makeOSKStringChangePacket(text);
+        sendPacketEncrypted(outPacket);
+    }
+
+    /**
+     * Tries to start an application on the PS4.
+     *
+     * @param applicationId The unique id for the application (CUSAxxxxx).
+     */
+    private void startApplication(String applicationId) {
+        ByteBuffer outPacket = PS4PacketHandler.makeApplicationPacket(applicationId);
+        sendPacketEncrypted(outPacket);
+    }
+
+    private void sendRemoteKey(int pushedKey) {
+        try {
+            SocketChannelHandler scHandler = socketChannelHandler;
+            int preWait = (scHandler == null || !loggedIn) ? POST_CONNECT_SENDKEY_DELAY_MS : 0;
+            SocketChannel channel = getConnection();
+
+            scheduler.schedule(() -> {
+                ByteBuffer outPacket = PS4PacketHandler.makeRemoteControlPacket(PS4_KEY_OPEN_RC);
+                sendPacketEncrypted(outPacket, channel);
+
+                scheduler.schedule(() -> {
+                    // Send remote key
+                    ByteBuffer keyPacket = PS4PacketHandler.makeRemoteControlPacket(pushedKey);
+                    sendPacketEncrypted(keyPacket, channel);
+
+                    scheduler.schedule(() -> {
+                        ByteBuffer offPacket = PS4PacketHandler.makeRemoteControlPacket(PS4_KEY_OFF);
+                        sendPacketEncrypted(offPacket, channel);
+
+                        scheduler.schedule(() -> {
+                            ByteBuffer closePacket = PS4PacketHandler.makeRemoteControlPacket(PS4_KEY_CLOSE_RC);
+                            sendPacketEncrypted(closePacket, channel);
+                        }, MIN_SENDKEY_DELAY_MS, TimeUnit.MILLISECONDS);
+
+                    }, MIN_HOLDKEY_DELAY_MS, TimeUnit.MILLISECONDS);
+
+                }, MIN_SENDKEY_DELAY_MS, TimeUnit.MILLISECONDS);
+
+            }, preWait, TimeUnit.MILLISECONDS);
```

Comment:
I'm sorry for the long delay in answering this, I just had to gather some energy.
Obviously I need to learn a bit more about openHAB and the ScheduledExecutorService.
I guess the Scheduler keeps some kind of reference to my binding to keep the GC in the VM from releasing the memory as long as it's scheduled, right?
And after openHAB has called dispose() I can't "talk" to openHAB any more, right?
Are there any more things to look out for in this context?

Maybe more to the point, what bad things can happen if a scheduler is not cancelled after dispose() is called?

Reply:
ScheduledExecutorService isn't really an openHAB thing. It is part of the core java concurrency utilities. The key thing you need to understand about ScheduledExecutorService is that you can use it to run task asynchronously in another thread. The ScheduledExecutorService manages a pool of threads that it uses to execute tasks that are submitted to it through the `execute` or `submit` methods. The `submit` methods return a `Future` instance which is tied to the task you submitted. Through that `Future` you wait and get the result of the task (if any) or cancel that task.

> I guess the Scheduler keeps some kind of reference to my binding to keep the GC in the VM from releasing the memory as long as it's scheduled, right?

That is true as long as the task is running, one all asynchronous tasks stop running the references can be cleanup with the GC since they are no longer in use.

> And after openHAB has called dispose() I can't "talk" to openHAB any more, right?

Not true, which is part of the problem. OpenHAB expects the handler to have been stopped after it calls `dispose`. There is nothing stopping the handler or binding from continuing to run and accessing openHAB resources even after `dispose` is called. But such a thing would cause a large number or problems which is why we enforce proper cleanup during `dispose`.

> Are there any more things to look out for in this context?

Asynchronous tasks can get tricky when it comes to cleanup and especially so when asynchronous tasks spawn other asynchronous tasks. My best advise it to try to minimize such behaviors if possible.

> Maybe more to the point, what bad things can happen if a scheduler is not cancelled after dispose() is called?

Well, the obvious problem is that the tasks keep running. If your task is meant to perform a periodic task, that periodic task would never stop. The only way to stop it at that point is to restart openHAB entirely.

But keep in mind, the `scheduler` that you inherit from the BaseThingHandler class doesn't require cleanup, only the Futures that you create with it require cleanup. Any executor service that you create yourself during `initialize` would require cleanup though.

Reply:
I absolutely understand that reoccurring tasks needs to be canceled, and that I can not / should not talk to openHAB after the plug-in is disposed.

But as the coding-guidelines for developing plug-ins states:
`Overridden methods from abstract classes or interfaces are expected to return fast unless otherwise stated in their JavaDoc. Expensive operations should therefore rather be scheduled as a job.`
I guess this includes the dispose method as well? And I need to send a bye bye message that somehow has 5 second timeout, how does this match up?
If this was clearer in the documentation it would have saved me a lot of time and effort, if this can be made clear in Coding Guidelines I guess it can save some other developers time and effort as well.

I really appreciate the time and effort you and the whole openHAB team has spent but I don't have enough of that right now.
So should I close this PR until I get some more of that, or just let it linger?

Reply:
For most protocols a bye message is not mandatory (in fact I don't know any binding where it has been absolutely necessary). The communication simply times out when OH disconnects and the other side drops the connection.

If I see correctly the nested scheduling in your code involves no periodic tasks and will finish deterministic. IMHO this is not as critical. But why don't you simply apply @cpmeister 's code suggestion, as it seems the last issue to merge this PR?

Reply:
The point is not about this specific issue with nested schedulers, I can probably just do them serially with higher delays. The point I'm trying find an answer to is if all schedulers need to be tracked all the time so they can be canceled in the dispose method and if you in that case can't use schedulers in the dispose method, even though you have to return "fast".

Reply:
All tasks with a delay or periodic tasks should be tracked and cancelled in `dispose()`. Starting tasks in `dispose()` should be avoided, but if there is a necessity we can discuss that.

Reply:
Ok. Are there any examples of such a tracker or guidelines on how to code one?

Reply:
> The general approach is to store a reference everytime you schedule a task and invoke `cancel()` on it in `dispose()`.
This works as long as you can be sure that the scheduler is not run again while one is still active, how does one make sure that is not the case?
 
> You could also add them to a list and iterate over the list in `dispose()`. But care must be taken that cancelled tasks are being removed to avoid memory leaks.
I guess all finished tasks must be removed as well? Otherwise you will just fill up the collection until dispose is called...

> An example can be found here: https://github.com/openhab/openhab-addons/pull/8634/files#diff-bbf0a7bd12dc4a35896e525a7330859ea9f746e9f895f7c4d1c9a1b76efacefbR202
Thanks I will take a look.